src/affine_dimensions.jl

```diff
@@ -0,0 +1,466 @@
+    return Dimensions{promote_type(R1,R2)}
+end
+function Base.promote_rule(::Type{SymbolicDimensions{R1}}, ::Type{AffineDimensions{R2}}) where {R1,R2}
+    return Dimensions{promote_type(R1,R2)}
+end
+function Base.promote_rule(::Type{AffineDimensions{R1}}, ::Type{SymbolicDimensions{R2}}) where {R1,R2}
+    return Dimensions{promote_type(R1,R2)}
+end
+
+
+
+
+# Conversions for Dimensions |> AffineDimenions =====================================================================================
+"""
+    uconvert(qout::UnionAbstractQuantity{<:Any, <:AbstractAffineDimensions}, q::UnionAbstractQuantity{<:Any, <:Dimensions})
+
+Convert a quantity `q` with base SI units to the affine units of `qout`, for `q` and `qout` with compatible units.
+You can also use `|>` as a shorthand for `uconvert`
+"""
+function uconvert(qout::UnionAbstractQuantity{<:Any, <:AffineDimensions}, q::UnionAbstractQuantity{<:Any, <:Dimensions})
+    @assert isone(ustrip(qout)) "You passed a quantity with a non-unit value to uconvert."
+    dout = dimension(qout)
+    dimension(q) == basedim(dout) || throw(DimensionError(q, qout))
+    vout = (ustrip(q)-offset(dout))/scale(dout)
+    return new_quantity(typeof(q), vout, dout)
+end
+
+function uconvert(qout::UnionAbstractQuantity{<:Any,<:AffineDimensions}, q::QuantityArray{<:Any,<:Any,<:Dimensions})
+    @assert isone(ustrip(qout)) "You passed a quantity with a non-unit value to uconvert."
+    dout = dimension(qout)
+    dimension(q) == basedim(dout) || throw(DimensionError(q, qout))
+    vout = (ustrip(q) .- offset(dout))./scale(dout)
+    return QuantityArray(vout, dout, quantity_type(q))
+end
+
+# Conversions for AbstractAffineDimensions |> AbstractSymbolicDimensions =======================================================
+function uconvert(qout::UnionAbstractQuantity{<:Any, <:AbstractSymbolicDimensions}, qin::AbstractQuantityOrArray{<:Any, <:AbstractAffineDimensions})
+    uconvert(qout, siunits(qin))
+end
+
+# Conversions for AbstractSymbolicDimensions |> AbstractAffineDimensions   =======================================================
+function uconvert(qout::UnionAbstractQuantity{<:Any,<:AbstractAffineDimensions}, qin::AbstractQuantityOrArray{<:Any, <:AbstractSymbolicDimensions})
+    uconvert(qout, siunits(qin))
+end
+
+# Conversions for AbstractAffineDimensions |> AbstractAffineDimensions =======================================================
+function uconvert(qout::UnionAbstractQuantity{<:Any, <:AbstractAffineDimensions}, qin::AbstractQuantityOrArray{<:Any, <:AbstractAffineDimensions})
+    uconvert(qout, siunits(qin))
+end
+
+# Multiplication and division of AffineDimensions ===============================================================
+function Base.:*(l::AffineDimensions, r::AffineDimensions) 
+    assert_no_offset(l)
+    assert_no_offset(r)
+    return AffineDimensions(
+        scale  = scale(l)*scale(r),
+        offset = offset(l),
+        basedim = basedim(l)*basedim(r)
+    )
+end
```

Comment:
Can these be automatically defined by overloading `map_dimensions`? This is what is done for symbolic dimensions